src/environmentd/tests/sql.rs

```diff
@@ -1291,6 +1291,132 @@ fn test_explain_timestamp_json() {
+
+    let server = util::start_server(config).unwrap();
+
+    let mut client_writes = server.connect(postgres::NoTls).unwrap();
+    let mut client_reads = server.connect(postgres::NoTls).unwrap();
+
+    client_writes
+        .batch_execute("CREATE TABLE t1 (i1 int)")
+        .unwrap();
+
+    // Verify execution during a write-only txn fails
+    client_writes.batch_execute("BEGIN").unwrap();
+    client_writes
+        .batch_execute("INSERT INTO t1 VALUES (1)")
+        .unwrap();
+    let error = client_writes
+        .query_one("EXPLAIN TIMESTAMP FOR SELECT * FROM t1;", &[])
+        .unwrap_err();
+
+    assert!(format!("{}", error).contains("transaction in write-only mode"));
+
+    client_writes.batch_execute("ROLLBACK").unwrap();
+
+    // Verify the transaction timestamp is returned for each select and
+    // read holds are acquired as of the transaction timestamp
+    client_reads.batch_execute("BEGIN").unwrap();
+    let mut query_timestamp = None;
+    let mut t1_read_frontier = None;
+
+    for i in 1..5 {
+        let row = client_reads
+            .query_one("EXPLAIN TIMESTAMP AS JSON FOR SELECT * FROM t1;", &[])
+            .unwrap();
+
+        let explain: String = row.get(0);
+        let explain: TimestampExplanation<Timestamp> = serde_json::from_str(&explain).unwrap();
+        let explain_timestamp = explain.determination.timestamp_context.timestamp().unwrap();
+
+        if let Some(timestamp) = query_timestamp {
+            assert_eq!(timestamp, *explain_timestamp);
+        } else {
+            query_timestamp = Some(*explain_timestamp);
+        }
+
+        let explain_t1_read_frontier = explain
+            .sources
+            .first()
+            .unwrap()
+            .read_frontier
+            .first()
+            .unwrap();
+
+        // Ensure `t1` does not undergo compaction
+        if let Some(timestamp) = t1_read_frontier {
+            assert_eq!(timestamp, *explain_t1_read_frontier);
+        } else {
+            t1_read_frontier = Some(explain_t1_read_frontier.clone());
+        }
+
+        // Increase now by 2s each iteration
+        *nowfn.lock().unwrap() = NowFn::from(move || 2000 * i);
+        // Inserting tends to cause sources to compact, so this should ideally
+        // strengthen the assertion above that `t1`'s read frontier should
+        // not advance during the txn
+        client_writes
+            .batch_execute("INSERT INTO t1 VALUES (1)")
+            .unwrap();
+    }
+
+    // Errors when an object outside the chosen time domain is referenced
+    let error = client_reads
+        .query_one(
+            "EXPLAIN TIMESTAMP FOR SELECT * FROM mz_catalog.mz_views;",
+            &[],
+        )
+        .unwrap_err();
+
+    assert!(format!("{}", error)
+        .contains("Transactions can only reference objects in the same timedomain"));
+
+    client_reads.batch_execute("COMMIT").unwrap();
+
+    client_reads.query_one("SELECT mz_now();", &[]).unwrap();
```

Comment:
What is this query for?

Reply:
Ah, I pulled that from here:

https://github.com/MaterializeInc/materialize/blob/f42f05b745a9ceaf0232219711f7f47f78af1d68/test/sqllogictest/timedomain.slt#L12-L17

 Updated with a comment here as well